File: smart-fridge-java-backend/src/main/java/css/ch/smartfridgejavabackend/user/services/UserServiceImpl.java
package css.ch.smartfridgejavabackend.user.services;

import css.ch.smartfridgejavabackend.auth.dto.SignupRequest;
import css.ch.smartfridgejavabackend.exceptions.UserAlreadyExistsException;
import css.ch.smartfridgejavabackend.user.dto.UserResponseDTO;
import css.ch.smartfridgejavabackend.user.mapper.UserMapper;
import css.ch.smartfridgejavabackend.user.model.User;
import css.ch.smartfridgejavabackend.user.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;


@Service
@RequiredArgsConstructor
public class UserServiceImpl implements UserService {
    private final UserRepository repo;
    private final UserMapper mapper;
    private final PasswordEncoder passwordEncoder;

    @Override
    public UserResponseDTO createUser(SignupRequest request) {
        if (repo.existsByEmail(request.email())) {
            throw new UserAlreadyExistsException("User mit email " + request.email() + " existiert bereits");
        }

        User user = mapper.map(request);
        user.setPassword(passwordEncoder.encode(request.password()));

        return mapper.map(repo.save(user));
    }

    @Override
    public UserResponseDTO getUserById(String userId) {
        return repo.findById(userId)
                .map(mapper::map)
                .orElseThrow(() -> new RuntimeException("User mit ID " + userId + " nicht gefunden"));
    }
}

File: smart-fridge-backend/src/main/kotlin/ch/bbzw/smartfridgebackend/user/service/UserServiceImpl.kt
package ch.bbzw.smartfridgebackend.user.service

import ch.bbzw.smartfridgebackend.auth.dto.SignupRequest
import ch.bbzw.smartfridgebackend.user.dto.UserResponse
import ch.bbzw.smartfridgebackend.exceptions.UserAlreadyExistsException
import ch.bbzw.smartfridgebackend.exceptions.UserNotFoundException
import ch.bbzw.smartfridgebackend.user.mapper.UserMapper
import ch.bbzw.smartfridgebackend.user.repository.UserRepository
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service

@Service
class UserServiceImpl(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder,
    private val mapper: UserMapper
) : UserService {

    override fun createUser(signupRequest: SignupRequest): UserResponse {
        if (userRepository.existsByEmail(signupRequest.email)) {
            throw UserAlreadyExistsException("User mit email ${signupRequest.email} existiert bereits")
        }

        var user = mapper.map(signupRequest)

        user = user.copy(
            password = passwordEncoder.encode(signupRequest.password)
        )

        return mapper.map(userRepository.save(user))
    }

    override fun getUserById(id: String): UserResponse {
        val user = userRepository.findById(id)
            .orElseThrow { UserNotFoundException("User not found with id $id") }

        return mapper.map(user)
    }
}
